mlir/lib/Analysis/Presburger/Coalesce.cpp

```diff
@@ -0,0 +1,660 @@
+        if (adjEqCase(basicSetVector, i, j, info_1, info_2, true)) {
+          i--;
+          break;
+        }
+      } else if (info_1.t &&
+                 info_2.cut.empty()) { // adj_eq Case for one equality
+        SmallVector<int64_t, 8> adjEq;
+        complement(info_1.t.getValue(), adjEq); 
+        if (adjEqCaseNoCut(basicSetVector, i, j, adjEq)) { // adj_eq noCut cas
+          i--;
+          break;
+        } else if (info_1.t && adjEqCase(basicSetVector, i, j, info_1, info_2,
+                                          false)) { // adjEq case
+          i--;
+          break;
+        }
+      } else if (info_2.t &&
+                 info_1.cut.empty()) { // adj_eq Case for one equality
+        SmallVector<int64_t, 8> adjEq;
+        complement(info_2.t.getValue(), adjEq); 
+        if (adjEqCaseNoCut(basicSetVector, j, i,
+                           adjEq)) { // adj_eq noCut cas
+          i--;
+          break;
+        } else if (info_2.t && adjEqCase(basicSetVector, j, i, info_2, info_1,
+                                          false)) { // adjEq case
+          i--;
+          break;
+        }
+      }
+    }
+  }
+  for (size_t i = 0; i < basicSetVector.size(); i++) {
+    new_set.addFlatAffineConstraints(basicSetVector[i]);
+  }
+  return new_set;
+}
+
+void addInequalities(FlatAffineConstraints &bs,
+                     const SmallVector<SmallVector<int64_t, 8>, 8> &inequalities) {
+  for (size_t k = 0; k < inequalities.size(); k++) {
+    bs.addInequality(inequalities[k]);
+  }
+}
+
+void addEqualities(FlatAffineConstraints &bs,
+                   const SmallVector<SmallVector<int64_t, 8>, 8> &equalities) {
+  for (size_t k = 0; k < equalities.size(); k++) {
+    bs.addEquality(equalities[k]);
+  }
+}
+
+bool adjIneqPureCase(SmallVector<FlatAffineConstraints, 4> &basicSetVector,
+                     unsigned i, unsigned j, Info &info_a, Info &info_b) {
+  FlatAffineConstraints newSet(basicSetVector[i].getNumIds(), basicSetVector[i].getNumSymbolIds());
+  addInequalities(newSet, info_a.redundant);
+  addInequalities(newSet, info_a.cut);
+  addInequalities(newSet, info_b.redundant);
+  addInequalities(newSet, info_b.cut);
+  addCoalescedBasicSet(basicSetVector, i, j, newSet);
+  return true;
+}
+
+void addCoalescedBasicSet(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    unsigned i, unsigned j, FlatAffineConstraints &bs) {
+  if (i < j) {
+    basicSetVector.erase(basicSetVector.begin() + j);
+    basicSetVector.erase(basicSetVector.begin() + i);
+  } else {
+    basicSetVector.erase(basicSetVector.begin() + i);
+    basicSetVector.erase(basicSetVector.begin() + j);
+  }
+  basicSetVector.push_back(bs);
+}
+
+bool protrusionCase(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    Info &info_a, Info &info_b, unsigned i, unsigned j) {
+  FlatAffineConstraints a = basicSetVector[i];
+  FlatAffineConstraints b = basicSetVector[j];
```

Comment:
As far as I can tell you only read from `a` and `b`, therefore these should be references.